vars/infra.groovy

```diff
@@ -103,6 +104,91 @@ boolean retrieveMavenSettingsFile(String settingsXml) {
   return false
 }
 
+/**
+ * Execute the body passed as closure with a Maven settings file using the
+ * Artifact Caching Proxy provider corresponding to the requested one defined
+ * via the agent's env.ARTIFACT_CACHING_PROXY variable, or 'azure' if not defined.
+ * This allows decreasing JFrog Artifactory bandwidth consumption, and increase reliability.
+ * There are currently three providers, one for each cloud used in Jenkins Infrastructure:
+ * "aws", "azure" and "do" (DigitalOcean).
+ * The available providers can be restricted by setting a global ARTIFACT_CACHING_PROXY_AVAILABLE_PROVIDERS
+ * variable on the Jenkins controller, with providers separated by a comma. Ex: 'aws,do' if the Azure provider is unavailable.
+ * A 'skip-artifact-caching-proxy' label can be added to pull request in order to punctually disable it.
+ */
+Object withArtifactCachingProxy(Closure body) {
+  boolean useArtifactCachingProxy = true
+
+  // As the env var ARTIFACT_CACHING_PROXY_PROVIDER can't be set on Azure VM agents,
+  // we're specifying a default provider if none is specified.
+  final String requestedProxyProvider = env.ARTIFACT_CACHING_PROXY_PROVIDER ?: 'azure'
+  final String[] validProxyProviders = ['aws', 'azure', 'do']
+  // Useful when a provider is in maintenance (or similar cases), add a global env var in Jenkins controller settings to restrict them.
+  // To completely disable the artifact caching proxies, this value can be set to a value absent of validProxyProviders like "none" for example.
+  final String configuredAvailableProxyProviders = env.ARTIFACT_CACHING_PROXY_AVAILABLE_PROVIDERS ?: validProxyProviders.join(',')
+  final String[] availableProxyProviders = configuredAvailableProxyProviders.split(',')
+
+  // Check if the requested provider is invalid or unavailable
+  if (!validProxyProviders.contains(requestedProxyProvider) || !availableProxyProviders.contains(requestedProxyProvider)) {
+    echo "WARNING: invalid or unavailable artifact caching proxy provider '${requestedProxyProvider}' requested by the agent, will use repo.jenkins-ci.org"
+    useArtifactCachingProxy = false
+  }
+
+  // If the build concerns a pull request, check if there is "skip-artifact-caching-proxy" label applied in case the user doesn't want ACP
+  if ((useArtifactCachingProxy) && (!env.BRANCH_IS_PRIMARY)) {
+    final String currentCredentialsId = SCMSource.SourceByItem.findSource(currentBuild.rawBuild.parent).credentialsId
```

Comment:
I do not recommend doing this except inside a dedicated `@NonCPS` method, since you do not want to try to serialize the `SCMSource` object inside the program data.

Reply:
just do this?
```
@NonCPS
def getBuildCredentialsId() {
  return SCMSource.SourceByItem.findSource(currentBuild.rawBuild.parent).credentialsId
}
```

Reply:
and then see if you can override it in https://github.com/jenkins-infra/pipeline-library/blob/07e4c9f5e5c000bfd2b995a381973e395d126b51/test/groovy/mock/Infra.groovy#L4